### PR TITLE
[TACHYON-1196] Enable FindBugs to prevent static circular dependencies

### PR DESCRIPTION
**Pull request by aaudiber**
https://tachyon.atlassian.net/browse/TACHYON-1196

On my local machine this added around 1.5 minutes to the build, but was able to detect the cyclic dependency issue we had recently (https://tachyon.atlassian.net/browse/TACHYON-1193)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -375,17 +375,33 @@
...
-      <!-- Find Bugs -->
+      <!-- FindBugs -->
       <plugin>
         <groupId>org.codehaus.mojo</groupId>
         <artifactId>findbugs-maven-plugin</artifactId>
-        <version>2.5.2</version>
+        <version>3.0.2</version>
         <configuration>
           <excludeFilterFile>${findbugs.path}findbugs-exclude.xml</excludeFilterFile>
+          <includeFilterFile>${findbugs.path}findbugs-include.xml</includeFilterFile>
```
Comment (yupeng9) [inline pom.xml]: With this change, do we still need findbugs.path configuration in each project's pom file?

Comment (aaudiber) [inline pom.xml]: I'm not sure why they were needed it in the first place, but it seems to work fine without them so they are now removed.

File: pom.xml [hunk 2]
```diff
@@ -375,17 +375,33 @@
...
+          <!--
+              Enables analysis which takes more memory but finds more bugs.
+              If you run out of memory, changes the value of the effort element
```
Comment (apc999) [inline pom.xml]: TODO: we should have a more detailed "how to build Tachyon" web page, covering different caveats

File: pom.xml (hunk 1 above)
Comment (apc999) [inline pom.xml]: NIT: is it a bit weird to have `${findbugs.path}findbugs-exclude.xml` rather than `${findbugs.path}/findbugs-exclude.xml`?

Comment (aaudiber) [inline pom.xml]: Agreed, fixed

**Comment by jsimsa** [comment]
Have you tried the approach that checks for circular dependencies using a unit test? I expect that the overhead of your solution is not fixed, but a function of the rest of the code base and would be interested in knowing how the unit test based solution compares to it.

**Comment by aaudiber** [comment]
I tried the JDepend unit test approach, but found that it will tell you that there is a cycle, but not where the cycle occurs. http://stackoverflow.com/questions/62276/java-package-cycle-detection-how-to-find-the-specific-classes-involved#71610

FindBugs is overkill for checking just circular dependencies, and the 1.5 minutes will increase as the code base grows. However, I will be un-filtering many new types of FindBugs warnings in the next few days, so the FindBugs value-add will grow and make it easier for us to review code.

**Comment by jsimsa** [comment]
SGTM

**Comment by yupeng9** [comment]
LGTM. Thanks for enabling Findbugs for such detection and look forward to more Findbugs rules to keep the codebase healthy.

File: pom.xml [hunk 3]
```diff
@@ -375,17 +375,33 @@
...
+          <effort>Max</effort>
+          <failOnError>true</failOnError>
+          <threshold>Low</threshold>
           <findbugsXmlOutput>true</findbugsXmlOutput>
           <xmlOutput>true</xmlOutput>
-          <effort>Max</effort>
         </configuration>
+        <executions>
+          <execution>
+            <phase>verify</phase>
```
Comment (calvinjia) [inline pom.xml]: Should this be run before tests?

Comment (aaudiber) [inline pom.xml]: Done, these are probably more likely to fail than the tests (once we unsuppress more warnings at least), and take less time.

**Comment by apc999** [comment]
+1 LGTM. Thanks for the patch!

**Pull request merged**